File: lib/rag.ts

import policies from './policies.json';

export interface Policy {
    id: string;
    title: string;
    text: string;
}

export function retrieveRelevantPolicies(receiptText: string): Policy[] {
    const text = receiptText.toLowerCase();

    // Simple keyword matching for Light RAG
    const relevant = policies.filter(p => {
        // Check various keywords associated with policies
        if (p.id === 'travel_3_2' && (text.includes('meal') || text.includes('food') || text.includes('restaurant') || text.includes('dinner') || text.includes('lunch') || text.includes('breakfast'))) return true;
        if (p.id === 'office_1_1' && (text.includes('paper') || text.includes('monitor') || text.includes('pen') || text.includes('notebook') || text.includes('supply') || text.includes('office'))) return true;
        if (p.id === 'transport_2_1' && (text.includes('uber') || text.includes('lyft') || text.includes('taxi') || text.includes('cab') || text.includes('ride') || text.includes('trip'))) return true;
        if (p.id === 'alcohol_policy' && (text.includes('wine') || text.includes('beer') || text.includes('alcohol') || text.includes('cocktail') || text.includes('bar'))) return true;

        // Default: if no specific keywords match, maybe include all or none? 
        // For "Light RAG", we want to be selective. 
        // Let's add a generic fallback if the text contains the policy title words
        return p.title.toLowerCase().split(' ').some(word => text.includes(word) && word.length > 3);
    });

    // If nothing matches, return a default set or empty (let's return all if none match to be safe, or just empty? 
    // The user prompt implies we MUST find relevant policies. Let's return all if empty as a fallback for the demo).
    return relevant.length > 0 ? relevant : policies;
}

File: python_backend/rag.py

import sys
import json
import os


def load_policies(filepath):
    """Load the company expense policies from a JSON file."""
    try:
        with open(filepath, 'r') as f:
            return json.load(f)
    except Exception:
        return []


def find_relevant_policies(text, policies):
    """
    Lightweight keyword-based retrieval (RAG without a vector DB).
    
    Matches receipt text against known policy keywords to retrieve
    only the relevant policies. This avoids sending the entire policy
    document to the LLM, reducing token usage and improving accuracy.
    """
    relevant = []
    text_lower = text.lower()

    # Keyword mapping for each policy category
    keywords = {
        "travel_3_2": ["meal", "food", "lunch", "dinner", "breakfast", "restaurant"],
        "office_1_1": ["office", "supplies", "stationery", "paper", "pen"],
        "transport_2_1": ["cab", "taxi", "uber", "lyft", "ride"],
        "alcohol_policy": ["alcohol", "wine", "beer", "liquor", "drink"]
    }

    for policy in policies:
        pid = policy.get("id")
        policy_keywords = keywords.get(pid, [])

        # Direct keyword match (high confidence)
        if any(k in text_lower for k in policy_keywords):
            relevant.append(policy)
            continue

        # Fuzzy word overlap match (lower confidence, needs 3+ common words)
        policy_words = set(policy.get("text", "").lower().split())
        text_words = set(text_lower.split())
        common = policy_words.intersection(text_words)
        if len(common) > 2:
            relevant.append(policy)

    return relevant


if __name__ == "__main__":
    if len(sys.argv) > 1:
        text_input = sys.argv[1]
    else:
        try:
            data = json.load(sys.stdin)
            text_input = data.get("text", "")
        except Exception:
            sys.exit(1)

    base_dir = os.path.dirname(os.path.abspath(__file__))
    policy_path = os.path.join(base_dir, "..", "lib", "policies.json")

    policies = load_policies(policy_path)
    relevant = find_relevant_policies(text_input, policies)

    print(json.dumps(relevant))
